source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.cc

```diff
@@ -0,0 +1,376 @@
+#include "source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.h"
+
+#include <memory>
+
+#include "envoy/upstream/cluster_manager.h"
+#include "envoy/upstream/thread_local_cluster.h"
+
+#include "source/common/common/utility.h"
+#include "source/extensions/filters/network/thrift_proxy/app_exception_impl.h"
+#include "source/extensions/filters/network/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+namespace Router {
+
+bool NullResponseDecoder::onData(Buffer::Instance& data) {
+  upstream_buffer_.move(data);
+
+  bool underflow = false;
+  decoder_->onData(upstream_buffer_, underflow);
+  ASSERT(complete_ || underflow);
+  return complete_;
+}
+
+FilterStatus NullResponseDecoder::messageBegin(MessageMetadataSharedPtr metadata) {
+  metadata_ = metadata;
+  first_reply_field_ =
+      (metadata->hasMessageType() && metadata->messageType() == MessageType::Reply);
+
+  return FilterStatus::Continue;
+}
+
+FilterStatus NullResponseDecoder::fieldBegin(absl::string_view, FieldType&, int16_t& field_id) {
+  if (first_reply_field_) {
+    // Reply messages contain a struct where field 0 is the call result and fields 1+ are
+    // exceptions, if defined. At most one field may be set. Therefore, the very first field we
+    // encounter in a reply is either field 0 (success) or not (IDL exception returned).
+    // If first fieldType is FieldType::Stop then it is a void success and handled in messageEnd()
+    // because decoder state machine does not call decoder event callback fieldBegin on
+    // FieldType::Stop.
+    success_ = (field_id == 0);
+    first_reply_field_ = false;
+  }
+
+  return FilterStatus::Continue;
+}
+
+FilterStatus NullResponseDecoder::messageEnd() {
+  if (first_reply_field_) {
+    // When the response is thrift void type there is never a fieldBegin call on a success
+    // because the response struct has no fields and so the first field type is FieldType::Stop.
+    // The decoder state machine handles FieldType::Stop by going immediately to structEnd,
+    // skipping fieldBegin callback. Therefore if we are still waiting for the first reply field
+    // at end of message then it is a void success.
+    success_ = true;
+    first_reply_field_ = false;
+  }
+
+  return FilterStatus::Continue;
+}
+
+FilterStatus NullResponseDecoder::transportEnd() {
+  ASSERT(metadata_ != nullptr);
+  complete_ = true;
+  // TODO: bump stats.
```

Comment:
Should this be part of this PR? If not, `// TODO(rgs1): ...`

Reply:
That precise comment might be out of date, but there are few other spots. I'll address them, assuming we generally agree with the direction of the PR.